File: Request-Reply/auto_client.py
import zmq
import sys
import os
import time
import random
import threading

sys.path.append('/app/common')

from common.message_protocol import MessageProtocol

class AutoBBSClient:
    def __init__(self, broker_host="broker", broker_port=5555, pubsub_host="pubsub-proxy", pubsub_port=5558):
        self.broker_url = f"tcp://{broker_host}:{broker_port}"
        self.pubsub_url = f"tcp://{pubsub_host}:{pubsub_port}"
        
        self.context = zmq.Context()
        self.req_socket = self.context.socket(zmq.REQ)
        self.req_socket.setsockopt(zmq.RCVTIMEO, 5000)
        self.req_socket.connect(self.broker_url)
        
        self.sub_socket = self.context.socket(zmq.SUB)
        self.sub_socket.connect(self.pubsub_url)
        
        self.username = f"bot_{random.randint(1000, 9999)}"
        self.receiving = False
        self.receive_thread = None
        
        # Mensagens pré-definidas para os bots
        self.messages = [
            "Olá pessoal!",
            "Como vocês estão?",
            "Alguém online?",
            "Que dia lindo!",
            "Alguma novidade?",
            "Estou testando o sistema",
            "Funcionando perfeitamente!",
            "Mensagem automática de teste",
            "Hello world!",
            "Sistema BBS é incrível!"
        ]
        
        # Canais disponíveis
        self.available_channels = ["geral", "tech", "random", "offtopic", "testes"]
    
    def send_request(self, message: str) -> dict:
        try:
            self.req_socket.send_string(message)
            response_str = self.req_socket.recv_string()
            return MessageProtocol.parse_message(response_str)
        except zmq.Again:
            return {"service": "error", "data": {"status": "erro", "description": "Timeout"}}
        except Exception as e:
            return {"service": "error", "data": {"status": "erro", "description": f"Erro: {str(e)}"}}
    
    def login(self):
        """Faz login com nome aleatório"""
        message = MessageProtocol.create_login_message(self.username)
        response = self.send_request(message)
        
        if response.get("service") == "login" and response["data"]["status"] == "sucesso":
            print(f"🤖 {self.username} conectado")
            
            # Inscrever no próprio usuário e em alguns canais
            self.sub_socket.setsockopt_string(zmq.SUBSCRIBE, f"user.{self.username}")
            for channel in random.sample(self.available_channels, 2):
                self.sub_socket.setsockopt_string(zmq.SUBSCRIBE, f"channel.{channel}")
                print(f"🤖 {self.username} inscrito em: {channel}")
            
            return True
        else:
            print(f"❌ {self.username} falhou no login")
            return False
    
    def ensure_channels_exist(self):
        """Garante que os canais existam"""
        for channel in self.available_channels:
            message = MessageProtocol.create_channel_message(channel)
            response = self.send_request(message)
            if response.get("service") == "channel":
                if response["data"]["status"] == "sucesso":
                    print(f"🤖 Canal criado: {channel}")
                # Se já existe, não faz nada
    
    def start_receiving(self):
        """Inicia thread para receber mensagens"""
        self.receiving = True
        self.receive_thread = threading.Thread(target=self._receive_loop, daemon=True)
        self.receive_thread.start()
    
    def stop_receiving(self):
        """Para de receber mensagens"""
        self.receiving = False
        if self.receive_thread:
            self.receive_thread.join(timeout=1.0)
    
    def _receive_loop(self):
        """Loop para receber mensagens"""
        while self.receiving:
            try:
                self.sub_socket.setsockopt(zmq.RCVTIMEO, 1000)
                message = self.sub_socket.recv_string()
                
                msg_data = MessageProtocol.parse_pubsub_message(message)
                sender = msg_data.get("sender", "Desconhecido")
                content = msg_data.get("content", "")
                target = msg_data.get("target", "")
                
                # Ignora próprias mensagens
                if sender != self.username:
                    if target.startswith("channel."):
                        channel = target.replace("channel.", "")
                        print(f"🤖 {self.username} recebeu em #{channel}: {sender}: {content}")
                    else:
                        print(f"🤖 {self.username} recebeu privado: {sender}: {content}")
                        
            except zmq.Again:
                continue
            except Exception as e:
                if self.receiving:
                    print(f"🤖 Erro ao receber: {e}")
    
    def send_random_message(self):
        """Envia uma mensagem aleatória"""
        channel = random.choice(self.available_channels)
        message = random.choice(self.messages)
        
        request = MessageProtocol.create_publish_message(self.username, channel, message)
        response = self.send_request(request)
        
        if response.get("service") == "publish" and response["data"]["status"] == "OK":
            print(f"🤖 {self.username} publicou em #{channel}: {message}")
        else:
            print(f"🤖 {self.username} falhou ao publicar")
    
    def run(self):
        """Loop principal do cliente automático"""
        print(f"🚀 Iniciando cliente automático: {self.username}")
        
        # Login e setup
        if not self.login():
            return
        
        self.ensure_channels_exist()
        self.start_receiving()
        
        try:
            message_count = 0
            while message_count < 10:  # Envia 10 mensagens
                # Espera entre 2 e 5 segundos
                wait_time = random.uniform(2.0, 5.0)
                time.sleep(wait_time)
                
                self.send_random_message()
                message_count += 1
            
            print(f"✅ {self.username} completou 10 mensagens. Continuando a ouvir...")
            
            # Continua ouvindo mensagens indefinidamente
            while True:
                time.sleep(10)
                
        except KeyboardInterrupt:
            print(f"\n🛑 {self.username} interrompido")
        finally:
            self.stop_receiving()
            print(f"👋 {self.username} finalizado")

if __name__ == "__main__":
    client = AutoBBSClient()
    client.run()

File: Request-Reply/c/auto_client.c
#include <zmq.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <msgpack.h>
#include <time.h>
#include <unistd.h>
#include <pthread.h>

#define BROKER_ENDPOINT "tcp://broker:5555"
#define PUBSUB_ENDPOINT "tcp://pubsub-proxy:5558"

// Incluir protocolo de mensagens
#include "common/logical_clock.h"
#include "common/message_protocol.h"

typedef struct {
    void* context;
    void* req_socket;
    void* sub_socket;
    char username[32];
    int message_count;
    int running;
    logical_clock_t clock; // Relógio lógico
} auto_client_t;



// Mensagens pré-definidas para os bots
const char* MESSAGES[] = {
    "Olá pessoal!",
    "Como vocês estão?",
    "Alguém online?",
    "Que dia lindo!",
    "Alguma novidade?",
    "Estou testando o sistema",
    "Funcionamento perfeito!",
    "Mensagem automática",
    "Hello world!",
    "Sistema BBS é ótimo!"
};
const int MESSAGE_COUNT = 10;

// Canais disponíveis
const char* CHANNELS[] = {"geral", "tech", "random", "offtopic", "testes"};
const int CHANNEL_COUNT = 5;

// Gerar nome de usuário aleatório
void generate_username(char* buffer, size_t size) {
    int random_num = rand() % 10000;
    snprintf(buffer, size, "bot_%04d", random_num);
}

// Função para enviar requisição usando MessagePack 
int send_request(auto_client_t* client, msgpack_sbuffer* request, msgpack_sbuffer* response) {
    // Incrementar relógio antes do envio
    logical_clock_increment(&client->clock);
    
    // Enviar mensagem
    int send_result = zmq_send(client->req_socket, request->data, request->size, 0);
    if (send_result == -1) {
        printf("❌ Erro ao enviar mensagem: %s\n", strerror(errno));
        return -1;
    }

    // Receber resposta com timeout de 5 segundos
    int timeout = 5000;  // 5000ms = 5s
    zmq_setsockopt(client->req_socket, ZMQ_RCVTIMEO, &timeout, sizeof(timeout));
    
    char reply_buffer[4096];
    int recv_result = zmq_recv(client->req_socket, reply_buffer, sizeof(reply_buffer) - 1, 0);
    if (recv_result == -1) {
        printf("❌ Erro ao receber resposta: %s\n", strerror(errno));
        return -1;
    }

    // Copiar dados MessagePack para buffer de resposta
    msgpack_sbuffer_clear(response);
    msgpack_sbuffer_write(response, reply_buffer, recv_result);
    
    // Atualizar relógio com resposta
    msgpack_object data;
    if (parse_message(response->data, response->size, &data) == 0) {
        int64_t received_clock = extract_clock_field(&data);
        if (received_clock > 0) {
            logical_clock_update(&client->clock, received_clock);
        }
    }
    
    printf("📥 Resposta recebida: %d bytes | ⏰ Clock: %ld\n", recv_result, logical_clock_get(&client->clock));
    return 0;
}

// Login do bot usando MessagePack
int bot_login(auto_client_t* client) {
    msgpack_sbuffer request_buf, response_buf;
    msgpack_sbuffer_init(&request_buf);
    msgpack_sbuffer_init(&response_buf);
    
    // Incluir clock na mensagem
    create_login_message(client->username, logical_clock_get(&client->clock), &request_buf);    
    if (send_request(client, &request_buf, &response_buf) != 0) {
        msgpack_sbuffer_destroy(&request_buf);
        msgpack_sbuffer_destroy(&response_buf);
        return -1;
    }
    
    // Parsear resposta serializada
    msgpack_object data;
    if (parse_message(response_buf.data, response_buf.size, &data) == 0) {
        char status[16] = "";
        char description[128] = "";
        
        extract_service_data(&data, status, sizeof(status), description, sizeof(description));
        
        if (strcmp(status, "sucesso") == 0) {
            printf("🤖 %s conectado | ⏰ Clock: %ld\n", client->username, logical_clock_get(&client->clock));
            
            // Inscrever em alguns canais aleatórios
            for (int i = 0; i < 2; i++) {
                const char* channel = CHANNELS[rand() % CHANNEL_COUNT];
                char topic[64];
                snprintf(topic, sizeof(topic), "channel.%s", channel);
                zmq_setsockopt(client->sub_socket, ZMQ_SUBSCRIBE, topic, strlen(topic));
                printf("🤖 %s inscrito em: %s\n", client->username, channel);
            }
            
            msgpack_sbuffer_destroy(&request_buf);
            msgpack_sbuffer_destroy(&response_buf);
            return 0;
        }
    }
    
    msgpack_sbuffer_destroy(&request_buf);
    msgpack_sbuffer_destroy(&response_buf);
    return -1;
}

// Garantir que os canais existam usando MessagePack
void ensure_channels_exist(auto_client_t* client) {
    msgpack_sbuffer request_buf, response_buf;
    
    for (int i = 0; i < CHANNEL_COUNT; i++) {
        msgpack_sbuffer_init(&request_buf);
        msgpack_sbuffer_init(&response_buf);
        
        // Incluir clock na mensagem
        create_channel_message(CHANNELS[i], logical_clock_get(&client->clock), &request_buf);
        
        if (send_request(client, &request_buf, &response_buf) == 0) {
            printf("🤖 Canal verificado: %s | ⏰ Clock: %ld\n", CHANNELS[i], logical_clock_get(&client->clock));
        }
        
        msgpack_sbuffer_destroy(&request_buf);
        msgpack_sbuffer_destroy(&response_buf);
    }
}

// Enviar mensagem aleatória usando MessagePack
void send_random_message(auto_client_t* client) {
    const char* channel = CHANNELS[rand() % CHANNEL_COUNT];
    const char* message = MESSAGES[rand() % MESSAGE_COUNT];
    
    msgpack_sbuffer request_buf, response_buf;
    msgpack_sbuffer_init(&request_buf);
    msgpack_sbuffer_init(&response_buf);
    
    // Incluir clock na mensagem
    create_publish_message(client->username, channel, message, logical_clock_get(&client->clock), &request_buf);
    
    if (send_request(client, &request_buf, &response_buf) == 0) {
        printf("🤖 %s publicou em #%s: %s | ⏰ Clock: %ld\n", 
               client->username, channel, message, logical_clock_get(&client->clock));
        client->message_count++;
        
        // Mostrar tamanho da mensagem MessagePack
        printf("   📦 Tamanho MessagePack: %zu bytes\n", request_buf.size);
    }
    
    msgpack_sbuffer_destroy(&request_buf);
    msgpack_sbuffer_destroy(&response_buf);
}

// Thread para receber mensagens usando MessagePack
void* receive_messages(void* arg) {
    auto_client_t* client = (auto_client_t*)arg;
    
    while (client->running) {
        char msg_buffer[4096];
        
        // Timeout de 1 segundo
        int timeout_ms = 1000;
        zmq_setsockopt(client->sub_socket, ZMQ_RCVTIMEO, &timeout_ms, sizeof(timeout_ms));
        
        int recv_result = zmq_recv(client->sub_socket, msg_buffer, sizeof(msg_buffer) - 1, 0);
        if (recv_result != -1) {
            // Incrementar relógio ao receber mensagem
            logical_clock_increment(&client->clock);
            
            // Parsear mensagem Pub/Sub serializada
            msgpack_object data;
            if (parse_pubsub_message(msg_buffer, recv_result, &data) == 0) {
                char sender[32] = "Desconhecido";
                char content[256] = "";
                char target[32] = "";
                
                extract_string_field(&data, "sender", sender, sizeof(sender));
                extract_string_field(&data, "content", content, sizeof(content));
                extract_string_field(&data, "target", target, sizeof(target));
                
                // Extrair e atualizar clock da mensagem recebida
                int64_t received_clock = extract_clock_field(&data);
                if (received_clock > 0) {
                    logical_clock_update(&client->clock, received_clock);
                }
                
                // Ignorar próprias mensagens
                if (strcmp(sender, client->username) != 0) {
                    if (strncmp(target, "channel.", 8) == 0) {
                        printf("🤖 %s recebeu em #%s: %s: %s | ⏰ Clock: %ld\n", 
                               client->username, target + 8, sender, content, logical_clock_get(&client->clock));
                    } else {
                        printf("🤖 %s recebeu privado: %s: %s | ⏰ Clock: %ld\n", 
                               client->username, sender, content, logical_clock_get(&client->clock));
                    }
                }
            } else {
                printf("❌ Erro ao parsear mensagem Pub/Sub\n");
            }
        }
    }
    
    return NULL;
}

// Loop principal do bot
void run_bot(auto_client_t* client) {
    printf("🚀 Iniciando cliente automático: %s\n", client->username);
    printf("📦 Usando MessagePack para serialização binária\n");
    printf("⏰ Relógio lógico implementado\n");
    
    // Inicializar relógio lógico
    logical_clock_init(&client->clock);
    
    // Conectar sockets
    client->context = zmq_ctx_new();
    if (!client->context) {
        printf("❌ %s: Erro ao criar contexto ZMQ\n", client->username);
        return;
    }
    
    client->req_socket = zmq_socket(client->context, ZMQ_REQ);
    client->sub_socket = zmq_socket(client->context, ZMQ_SUB);
    
    if (!client->req_socket || !client->sub_socket) {
        printf("❌ %s: Erro ao criar sockets\n", client->username);
        goto cleanup;
    }
    
    // Configurar timeout para requisições
    int timeout_ms = 5000; // 5 segundos
    zmq_setsockopt(client->req_socket, ZMQ_RCVTIMEO, &timeout_ms, sizeof(timeout_ms));
    
    if (zmq_connect(client->req_socket, BROKER_ENDPOINT) != 0) {
        printf("❌ %s: Erro ao conectar com broker\n", client->username);
        goto cleanup;
    }
    
    if (zmq_connect(client->sub_socket, PUBSUB_ENDPOINT) != 0) {
        printf("❌ %s: Erro ao conectar com pubsub\n", client->username);
        goto cleanup;
    }
    
    printf("✅ %s: Conectado ao sistema BBS | ⏰ Clock: %ld\n", client->username, logical_clock_get(&client->clock));
    
    // Login
    if (bot_login(client) != 0) {
        printf("❌ %s: Falha no login\n", client->username);
        goto cleanup;
    }
    
    // Garantir canais
    ensure_channels_exist(client);
    
    // Iniciar thread de recebimento
    client->running = 1;
    pthread_t receive_thread;
    if (pthread_create(&receive_thread, NULL, receive_messages, client) != 0) {
        printf("❌ %s: Erro ao criar thread de recebimento\n", client->username);
        goto cleanup;
    }
    
    // Loop de envio de mensagens
    int messages_sent = 0;
    while (messages_sent < 10 && client->running) {
        // Espera entre 2 e 5 segundos
        int wait_time = 2 + (rand() % 4);
        sleep(wait_time);
        
        send_random_message(client);
        messages_sent++;
    }
    
    printf("✅ %s completou %d mensagens. Continuando a ouvir... | ⏰ Clock: %ld\n", 
           client->username, messages_sent, logical_clock_get(&client->clock));
    
    // Manter recebendo mensagens
    while (client->running) {
        sleep(10);
    }
    
    // Aguardar thread terminar
    pthread_join(receive_thread, NULL);

cleanup:
    // Limpeza
    client->running = 0;
    
    if (client->req_socket) {
        zmq_close(client->req_socket);
    }
    if (client->sub_socket) {
        zmq_close(client->sub_socket);
    }
    if (client->context) {
        zmq_ctx_destroy(client->context);
    }
    
    printf("👋 %s finalizado | ⏰ Clock final: %ld\n", client->username, logical_clock_get(&client->clock));
}

int main(int argc, char* argv[]) {
    printf("🤖 Iniciando Cliente Automático BBS em C\n");
    printf("📦 Serialização: MessagePack (formato binário)\n");
    printf("⏰ Parte 4: Relógios Lógicos implementados\n");
    
    auto_client_t client;
    memset(&client, 0, sizeof(client));
    
    // Seed o gerador de números aleatórios ANTES de gerar username
    // Usar /dev/urandom para melhor aleatoriedade
    FILE* urandom = fopen("/dev/urandom", "r");
    unsigned int seed;
    if (urandom) {
        fread(&seed, sizeof(seed), 1, urandom);
        fclose(urandom);
    } else {
        seed = time(NULL) ^ getpid() ^ clock();
    }
    srand(seed);
    
    // Gerar username único
    generate_username(client.username, sizeof(client.username));
    
    // Executar bot
    run_bot(&client);
    
    return 0;
}